docs/porting-guide.rst

```diff
@@ -1,6 +1,115 @@
+variable which is unknown to the GC but contains a pointer to a GC-managed
+object, the variable will point to invalid memory as soon as the object is
+moved.
+
+Back to ``HPy`` vs ``HPyField``:
+
+  * ``HPy`` handles must be used for all C local variables, function arguments
+    and function return values. They are supposed to be short-lived and closed
+    as soon as they are no longer needed. The debug mode will report a
+    long-lived ``HPy`` as a potential memory leak.
+
+  * In PyPy, they are implemented using an indirection: they are indexes
+    inside a big list of GC-managed objects: this big list is tracked by the
+    GC, so when an object move its pointer is correctly updated.
+
+  * ``HPyField`` is for long-lived references, and the GC must be aware of
+    their location in memory. In PyPy, an ``HPyField`` is implemented as a
+    direct pointer to the object, and thus we need a way to inform the GC
+    where it is in memory, so that it can update its value upon moving: this
+    job is done by ``tp_traverse``, as explained in the next section.
+
+  * On CPython, both ``HPy`` and ``HPyField`` are implemented as ``PyObject *``.
+
+**IMPORTANT**: if you write a custom type having ``HPyField`` s, you **MUST**
+also write a ``tp_traverse`` slot. Note that this is different than the old
+Python/C API, where you need ``tp_traverse`` only under certain
+conditions. See the next section for more details.
+
+``tp_traverse``, ``tp_clear``, ``Py_TPFLAGS_HAVE_GC``
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Let's quote the Python/C documentation about `GC support
+<https://docs.python.org/3/c-api/gcsupport.html>`_
+
+  Python’s support for detecting and collecting garbage which involves
+  circular references requires support from object types which are
+  “containers” for other objects which may also be containers. Types which do
+  not store references to other objects, or which only store references to
+  atomic types (such as numbers or strings), do not need to provide any
+  explicit support for garbage collection.
+
+A good rule of thumb is that if your type contains ``PyObject *`` fields, you
+need to:
+
+  1. provide a ``tp_traverse`` slot;
+
+  2. provide a ``tp_clear`` slot;
+
+  3. add the ``Py_TPFLAGS_GC`` to the ``tp_flags``.
+
+
+However, if you know that your ``PyObject *`` fields will contain only
+"atomic" types, you can avoid these steps.
+
+In HPy the rules are slightly different:
+
+  1. if you have a field of type ``HPyField``, you always **MUST** provide a
+     ``tp_traverse``. This is needed so that a moving GC can track the
+     relevant areas of memory;
+
+  2. ``tp_clear`` does not exist. On CPython, ``HPy`` automatically generates
+     one for you, by using ``tp_traverse`` to know which are the fields to
+     clear. Other implementations are free to ignore it, if it's not needed;
+
+  3. ``HPy_TPFLAGS_GC`` is still needed, especially on CPython. If you don't
```

Comment:
Could we do away with ``HPy_TPFLAGS_GC`` entirely and just use the presence to `tp_traverse` to indicate that the object has `HPyField`s to track? On CPython we can then set ``Py_TPFLAGS_GC`` based on `tp_traverse` being set.

Perhaps this simplification has some backwards compatibility issues? E.g. code in extensions that wants to check whether ``HPy_TPFLAGS_GC`` is set?

Reply:
we are discussing it right now. The problem is:
1. `Py_TPFLAGS_GC` is costly on CPython, because you need to allocate extra space for the GC header
2. on HPy you will need to add `tp_traverse` also for objects which currently don't need it on CPython

So, if we automatically add the flag based on the presence of `tp_traverse`, we end up with cases in which you pay a penalty which you don't need.

One option is to do the opposite: use the flag by default and add another flag to disable it, as an optimization. But it looks a bit obscure, especially for people who are already used to the current behavior.